src/surface_tracker/location.py

```diff
@@ -298,15 +303,63 @@ def __map_points(
 ##### Helper functions
 
 
-def _perspective_transform(points, transform_matrix):
-    homogeneous_points = cv2.convertPointsToHomogeneous(points)
+def is_clockwise_triangle(points):
+    p1, p2, p3 = points
+    val = (float(p2[1] - p1[1]) * (p3[0] - p2[0])) - (
+        float(p2[0] - p1[0]) * (p3[1] - p2[1])
+    )
+    return val < 0
+
+
+def orientation_quadrupel(points, clockwise=True):
+    p0, p1, p2, p3 = points
+    return all(
+        (
+            is_clockwise_triangle((p0, p1, p2)) == clockwise,
+            is_clockwise_triangle((p1, p2, p3)) == clockwise,
+            is_clockwise_triangle((p2, p3, p0)) == clockwise,
+        )
+    )
+
+
+def project_points_pos_z(homogeneous_points, transform_matrix):
```

Comment:
Also here, I think you can get rid of the for loop. Probably known to you, but for an (n,m) array of points you can calculate the row-wise norm by:
`np.linalg.norm(points, axis=1, keepdims=True)`

Reply:
I don't see how this applies here, as only the norm of the x and y coordinates is computed if the z value is negative. (This ensures that points lying in negative z direction are not mirrored at the origin, when converting from homogeneous coordinates to euclidean space, and at the same time are far away and not visible in the image.)

Reply:
Yeah, I see. I thought of something like this:
```
homogeneous_points = cv2.convertPointsToHomogeneous(points)[:,0,:]
projected_points = (transform_matrix @ homogeneous_points.T).T
idx = projected_points[:,2]<0
temp = np.vstack((-projected_points[idx,2], np.linalg.norm(projected_points[idx,:2], axis=1))
projected_points[idx, 2] = np.min(temp, axis=0)
```
But then I agree that does not make it more readable and speed is not an issue here I suppose.